pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/json/OffHeapJsonIndexCreator.java

```diff
@@ -0,0 +1,279 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.creator.impl.inv.json;
+
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import it.unimi.dsi.fastutil.longs.LongList;
+import java.io.BufferedOutputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import org.apache.pinot.core.io.util.VarLengthValueWriter;
+import org.apache.pinot.core.segment.creator.impl.inv.BitmapInvertedIndexWriter;
+import org.apache.pinot.core.segment.memory.PinotDataBuffer;
+import org.apache.pinot.spi.utils.StringUtils;
+import org.roaringbitmap.RoaringBitmap;
+import org.roaringbitmap.RoaringBitmapWriter;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+/**
+ * Implementation of {@link org.apache.pinot.core.segment.creator.JsonIndexCreator} that uses off-heap memory.
+ * <p>The posting lists (map from value to doc ids) are initially stored in a TreeMap, then flushed into a file for
+ * every 50000 documents (unflattened records) added. After all the documents are added, we read all the posting lists
+ * from the file and merge them using a priority queue to calculate the final posting lists. Then we generate the string
+ * dictionary and inverted index from the final posting lists and create the json index on top of them.
+ */
+public class OffHeapJsonIndexCreator extends BaseJsonIndexCreator {
+  private static final int FLUSH_THRESHOLD = 50_000;
```

Comment:
This should be a function of total doc size, not just num docs?

Reply:
why do you think so?

Reply:
We limit the on-heap memory usage by putting a threshold on the docs held in memory. The total doc size doesn't matter because it just increases the number of chunks, and won't increase the on-heap memory usage.